### PR TITLE
handle an OAUTH2 offline token refresh

### PR DESCRIPTION
**Pull request by lpardoRH**
This PR adds OAUTH2 offline token handle for ZAP users. 

* Authentication/OfflineTokenRefresher.js will automatically fetch the new access token for every unauthorized request determined by the "Logged Out" or "Logged In" indicator previously set in Context -> Authentication.

* HTTP Sender/AddBearerTokenHeader.js will add the new access token to all requests in scope made by ZAP (except the authentication ones) as an "Authorization: Bearer [access_token]" HTTP Header.

### CODE REVIEW

**Comment by lpardoRH** [comment]
any updates on this? Am I missing a previous step or anything for this to get reviewed/accepted?

File: authentication/OfflineTokenRefresh.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
+/*
+* This script is intended to be used along with HTTP Sender/AddBearerTokenHeader.js to
```
Comment (kingthorin) [inline authentication/OfflineTokenRefresh.js:2]:
```suggestion
* This script is intended to be used along with  httpsender/AddBearerTokenHeader.js  to
```

File: authentication/OfflineTokenRefresh.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+* handle an OAUTH2 offline token refresh workflow.
+*
+* Authentication/OfflineTokenRefresher.js will automatically fetch the new access token for every unauthorized
```
Comment (kingthorin) [inline authentication/OfflineTokenRefresh.js:5]:
```suggestion
* authentication/OfflineTokenRefresher.js will automatically fetch the new access token for every unauthorized
```

File: authentication/OfflineTokenRefresh.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+* request determined by the "Logged Out" or "Logged In" indicator previously set in Context -> Authentication.
+*
+* HTTP Sender/AddBearerTokenHeader.js will add the new access token to all requests in scope
```
Comment (kingthorin) [inline authentication/OfflineTokenRefresh.js:8]:
```suggestion
*  httpsender/AddBearerTokenHeader.js  will add the new access token to all requests in scope
```

File: authentication/OfflineTokenRefresh.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+* made by ZAP (except the authentication ones) as an "Authorization: Bearer [access_token]" HTTP Header.
+*
```
Comment (kingthorin) [inline authentication/OfflineTokenRefresh.js:10]:
```suggestion
```

File: authentication/OfflineTokenRefresh.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+*
+* @author Laura Pardo <lpardo at redhat.com>
+*
```
Comment (kingthorin) [inline authentication/OfflineTokenRefresh.js:13]:
```suggestion
```

File: httpsender/AddBearerTokenHeader.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
+/*
+* This script is intended to be used along with Authentication/OfflineTokenRefresher.js to
+* handle an OAUTH2 offline token refresh workflow.
+*
+* Authentication/OfflineTokenRefresher.js will automatically fetch the new access token for every unauthorized
+* request determined by the "Logged Out" or "Logged In" indicator previously set in Context -> Authentication.
+*
+* HTTP Sender/AddBearerTokenHeader.js will add the new access token to all requests in scope
```
Comment (kingthorin) [inline httpsender/AddBearerTokenHeader.js:8]:
```suggestion
*  httpsender/AddBearerTokenHeader.js will add the new access token to all requests in scope
```

File: httpsender/AddBearerTokenHeader.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
+/*
+* This script is intended to be used along with Authentication/OfflineTokenRefresher.js to
+* handle an OAUTH2 offline token refresh workflow.
+*
+* Authentication/OfflineTokenRefresher.js will automatically fetch the new access token for every unauthorized
```
Comment (kingthorin) [inline httpsender/AddBearerTokenHeader.js:5]:
```suggestion
* authentication/OfflineTokenRefresher.js will automatically fetch the new access token for every unauthorized
```

File: httpsender/AddBearerTokenHeader.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
+/*
+* This script is intended to be used along with Authentication/OfflineTokenRefresher.js to
```
Comment (kingthorin) [inline httpsender/AddBearerTokenHeader.js:2]:
```suggestion
* This script is intended to be used along with authentication/OfflineTokenRefresher.js to
```

File: httpsender/AddBearerTokenHeader.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+* handle an OAUTH2 offline token refresh workflow.
+*
+* Authentication/OfflineTokenRefresher.js will automatically fetch the new access token for every unauthorized
+* request determined by the "Logged Out" or "Logged In" indicator previously set in Context -> Authentication.
+*
+* HTTP Sender/AddBearerTokenHeader.js will add the new access token to all requests in scope
+* made by ZAP (except the authentication ones) as an "Authorization: Bearer [access_token]" HTTP Header.
+*
+*
```
Comment (kingthorin) [inline httpsender/AddBearerTokenHeader.js:11]:
```suggestion
```

File: httpsender/AddBearerTokenHeader.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+* @author Laura Pardo <lpardo at redhat.com>
+*
```
Comment (kingthorin) [inline httpsender/AddBearerTokenHeader.js:13]:
```suggestion
```

**Review by kingthorin — CHANGES_REQUESTED** [review]
Just minor things for consistency

**Comment by kingthorin** [comment]
The conflict in CHANGELOG.md also needs to be addressed.

**Comment by kingthorin** [comment]
To address the DCO requirement you'll need to sign-off the commit(s):

- https://github.com/zaproxy/zaproxy/blob/develop/CONTRIBUTING.md#developer-certificate-of-origin
- https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff

File: CHANGELOG.md [hunk 11]
```diff
@@ -14,6 +14,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 - httpfuzzerprocessor/unexpected_responses.js > compare response codes to a (pass/fail) regex and generate alerts
 - targeted/dns-email-spoofing > Check if DMARC / SPF policies are configured on a domain.
 - httpsender/add-more-headers.js > Add caller-specified headers to all requests.
+- authentication/OfflineTokenRefresh.js - refresh oauth2 offline tokens
+- httpsender/AddBearerTokenHeader.js - refresh oauth2 offline tokens
```
Comment (kingthorin) [inline CHANGELOG.md:18]: These should be moved into the Unreleased section:

```md
## [Unreleased]
### Added
- authentication/OfflineTokenRefresh.js - refresh oauth2 offline tokens
- httpsender/AddBearerTokenHeader.js - refresh oauth2 offline tokens
- ```

Comment (lpardoRH) [reply CHANGELOG.md]: changed, I probably merged it in the wrong place without noticing it

Comment (kingthorin) [reply CHANGELOG.md]: You got it in the right spot but missed the `### Added`

**Comment by kingthorin** [comment]
To address the DCO requirement you'll need to sign-off the commit(s):

- https://github.com/zaproxy/zaproxy/blob/develop/CONTRIBUTING.md#developer-certificate-of-origin
- https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff

**Review by kingthorin — CHANGES_REQUESTED** [review]
Getting close

**Review by kingthorin — APPROVED** [review]
Looks good to me

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Pull request merged**